### PR TITLE
Remove cmd exec cargo culting

### PR DESCRIPTION
**Pull request by busterb**
It seems like the cmd_exec interface has some weird quirks and interesting features, or at least used to, so modules have worked around them for years. But are they still actually an issue? I argue no.

This PR normalizes cmd_exec to actually imply that it works consistently across all session types, then takes a first stab at normalizing post/osx/gather/enum_osx do use the actual API rather than trying to workaround issues at the module level. This should simplify things a lot, and hopefully not kill any actual edge-case handling in the process. It also removes a lot of unnecessary 'chomp' calls, since cmd_exec already does this as well.

Tested lightly with osx meterpreter so far. See https://github.com/rapid7/metasploit-framework/issues/10046#issuecomment-389849775 for some recent discussion and links to other background.

### CODE REVIEW

File: modules/post/osx/gather/enum_osx.rb [hunk 1]
```diff
@@ -176,35 +153,18 @@ def enum_conf(log_folder)
...
-          file_local_write(log_folder+"//#{name}.txt",returned_data)
-        elsif session.type =~ /shell/
-          begin
-            returned_data = cmd_exec("/usr/sbin/system_profiler #{profile_datatypes}")
-            # Save data lo log folder
-            file_local_write(log_folder+"//#{name}.txt",returned_data)
-          rescue
-          end
-        end
+      returned_data = cmd_exec("/usr/sbin/system_profiler #{profile_datatypes}")
```
Comment (bcoles) [inline modules/post/osx/gather/enum_osx.rb]: The module exclusively uses `cmd_exec(cmd, arg)` format with two arguments, where as in this instance it's using `cmd_exec(cmd arg)` with one argument.

Not an issue, just inconsistent.

File: modules/post/osx/gather/enum_osx.rb [hunk 2]
```diff
@@ -459,7 +386,7 @@ def get_keychains(log_folder)
         end
       end
     else
-      current_user = cmd_exec("/usr/bin/id -nu").chomp
+      current_user = cmd_exec("/usr/bin/id -nu")
       print_status("Enumerating and Downloading keychains for #{current_user}")
       keychain_files = cmd_exec("usr/bin/security list-keychains").split("\n")
```
Comment (bcoles) [inline modules/post/osx/gather/enum_osx.rb]: Not related to this PR, but `usr/bin/security list-keychains` should almost certainly be prefaced with `/`

File: modules/post/osx/gather/enum_osx.rb [hunk 3]
```diff
@@ -105,42 +105,19 @@ def log_folder_create(log_path = nil)
...
   def check_server
     # Get the OS Name
-    case session.type
-    when /meterpreter/
-      osx_ver = cmd_exec("/usr/bin/sw_vers", "-productName").chomp
-    when /shell/
-      osx_ver = cmd_exec("/usr/bin/sw_vers -productName").chomp
-    end
-    return osx_ver =~/Server/
+    cmd_exec("/usr/bin/sw_vers", "-productName") =~/Server/
```
Comment (bcoles) [inline modules/post/osx/gather/enum_osx.rb]: Regex is unnecessary here. `include?` would be cleaner.

`cmd_exec("/usr/bin/sw_vers", "-productName").to_s.include? 'Server'`

**Comment by bcoles** [comment]
I gave this a once-over and it looks good to me. All of my comments are unrelated to the PR, and are instead related to the existing module code, or unrelated potential future enhancements to the framework.

---

However, I have some additional concerns surrounding the enum_osx module.

Some of the `file_local_write` calls make me nervous. Compounding this, the code used to list files is based on unsafe assumptions.

Using `ls -ma` then splitting the output by `', '` does not take into account new lines. This is bad. This code pattern is used extensively:

```
# grep -rn " -ma" modules/post/osx/gather/enum_osx.rb
263:        home_folder_list = cmd_exec("/bin/ls -ma ~/").chomp.split(", ")
268:          ssh_folder = cmd_exec("/bin/ls -ma ~/.ssh").chomp.split(", ")
282:          gnugpg_folder = cmd_exec("/bin/ls -ma ~/.gnupg").chomp.split(", ")
306:          user_folder = cmd_exec("/bin/ls -ma /Users/#{u}/").chomp.split(", ")
309:            ssh_folder = cmd_exec("/bin/ls -ma /Users/#{u}/.ssh").chomp.split(", ")
323:          user_folder = cmd_exec("/bin/ls -ma /Users/#{u}/").chomp.split(", ")
326:            ssh_folder = cmd_exec("/bin/ls -ma /Users/#{u}/.gnupg").chomp.split(", ")
392:      root_folder = cmd_exec("/bin/ls -ma ~/").chomp.split(", ")
408:        user_folder = cmd_exec("/bin/ls -ma /Users/#{u}/").chomp.split(", ")
422:      current_user_folder = cmd_exec("/bin/ls -ma ~/").chomp.split(", ")
```

Observe:

```
root@kali:/pentest/exploit/metasploit-framework# cat test.rb 
#!/usr/bin/env ruby

root_folder = `/bin/ls -ma ~/`.chomp.split(", ")
root_folder.each do |f|
  if f =~ /\.\w*\_history/
    puts "\tHistory file #{f.strip} found for root"
  end
end

root@kali:/pentest/exploit/metasploit-framework# ./test.rb 
	History file .bash_history found for root
	History file msfrpc-client,
.mysql_history found for root
	History file .pry_history found for root
	History file .psql_history found for root
	History file .rsf_history found for root
	History file .scapy_history,
.selenium found for root
	History file .sqlite_history found for root
```

**Comment by bcoles** [comment]
Works for me.

```
msf5 exploit(multi/handler) > use post/osx/gather/enum_osx 
msf5 post(osx/gather/enum_osx) > set session 1
session => 1
msf5 post(osx/gather/enum_osx) > run

[*] Running module against qqs-MacBook.local
[*] Saving all data to /root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123
[*] 	Enumerating OS
[*] 	Enumerating Network
[*] 	Enumerating Bluetooth
[*] 	Enumerating Ethernet
[*] 	Enumerating Printers
[*] 	Enumerating USB
[*] 	Enumerating Airport
[*] 	Enumerating Firewall
[*] 	Enumerating Known Networks
[*] 	Enumerating Applications
[*] 	Enumerating Development Tools
[*] 	Enumerating Frameworks
[*] 	Enumerating Logs
[*] 	Enumerating Preference Panes
[*] 	Enumerating StartUp
[*] 	Enumerating TCP Connections
[*] 	Enumerating UDP Connections
[*] 	Enumerating Environment Variables
[*] 	Enumerating Last Boottime
[*] 	Enumerating Current Activity
[*] 	Enumerating Process List
[*] 	Enumerating Users
[*] 	Enumerating Groups
[*] Capturing screenshot
[*] Screenshot Captured
[*] Extracting history files
[*] 	History file .bash_history found for qq
[*] 	Downloading .bash_history
[*] Enumerating and Downloading keychains for qq
[*] Post module execution completed
msf5 post(osx/gather/enum_osx) > ls /root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123
[*] exec: ls /root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123

Airport.txt
Applications.txt
Bluetooth.txt
Current Activity.txt
Development Tools.txt
Environment Variables.txt
Ethernet.txt
Firewall.txt
Frameworks.txt
Groups.txt
Known Networks.txt
Last Boottime.txt
Logs.txt
Network.txt
OS.txt
Preference Panes.txt
Printers.txt
Process List.txt
qq_.bash_history.txt
qq__line_86__usr_bin_security__No_such_file_or_directory
screenshot.jpg
StartUp.txt
TCP Connections.txt
UDP Connections.txt
USB.txt
Users.txt
msf5 post(osx/gather/enum_osx) > 
msf5 post(osx/gather/enum_osx) > cat /root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123/qq__line_86__usr_bin_security__No_such_file_or_directory
[*] exec: cat /root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123/qq__line_86__usr_bin_security__No_such_file_or_directory

cat: :: No such file or directory
cat: line: No such file or directory
cat: 86:: No such file or directory
cat: usr/bin/security:: No such file or directory
cat: No: No such file or directory
cat: such: No such file or directory
cat: file: No such file or directory
cat: or: No such file or directory
cat: directory: No such file or directory
```

Note that there was an an error resulting in an erroneous filename - `/root/.msf4/logs/post/enum_osx/qqs-MacBook.local_20180522.5123/qq__line_86__usr_bin_security__No_such_file_or_directory` - generated by the `get_keychains` method.

This is one of many issues with the `enum_osx` module and existed before this PR.

**Comment by bcoles** [comment]
@busterb I think we're good to land here.

I'll arrange resolving the issues in the `enum_osx` module in a separate issue/PR.

**Comment by wvu-r7** [comment]
@bcoles: All yours! LGTM.

**Comment by busterb** [comment]
Thanks guys.

File: modules/post/osx/gather/enum_osx.rb [hunk 4]
```diff
@@ -105,42 +105,19 @@ def log_folder_create(log_path = nil)
...
-    when /meterpreter/
-      id = cmd_exec("/usr/bin/id", "-ru").chomp
-    end
-
-    if id == "0"
-      return true
-    else
-      return false
-    end
+    cmd_exec("/usr/bin/id", "-ru") == "0"
```
[the remark this answers is not in the document]
Comment (busterb) [reply modules/post/osx/gather/enum_osx.rb]: Totally agree.

File: modules/post/osx/gather/enum_osx.rb [hunk 5]
```diff
@@ -30,7 +30,7 @@ def run
     when /meterpreter/
       host = sysinfo["Computer"]
     when /shell/
-      host = cmd_exec("hostname").chomp
+      host = cmd_exec("hostname")
```
[the remark this answers is not in the document]
Comment (busterb) [reply modules/post/osx/gather/enum_osx.rb]: I had a very similar thought.

**Comment by busterb** [comment]
## Release Notes

This reworks the cmd_exec post exploitation API to be more consistent between Meterpreter and Shell sessions. Doing so reduces the complexity of modules, and makes it easier to get consistent results regardless of the session type.

**Pull request merged**